### PR TITLE
Mature Builder logic for selecting Seals for inclusion in the block

### PR DESCRIPTION
**Pull request by AlexHentschel**
implements https://github.com/dapperlabs/flow-go/issues/5398

### Overview:
* PR https://github.com/onflow/flow-go/pull/462/ separates Receipts from Results in the block payload. This allows to streamline the sealing logic (in the `Builder` as well as the `sealValidator`)
* Fixed unhandled edge-case in `Builder`: Specifically, the implementation now handles the situation, when there are multiple candidate seals for the same block:
   * implement business logic 
   * add unit test
* Added missing check to `sealValidator`: check that the sealed execution results form a chain
   * implement business logic 
   * add unit test
* Extended documentation, including detailed comments on the correctness of the code
* update block traversal to use `TraverseBackward` and `TraverseForward`

#### Comment on Block Traversal

Initially, I used the `state.TraverseBackward` and `state.TraverseForward` methods for the builder and seal validator. But I was running into some very subtle bugs. The main problem is that the `state.TraverseBackward` always visits the provided start block. However, this is undesired in case we are building on the root block: https://github.com/onflow/flow-go/blob/32a1b71bd610717d439ce74151f76f5b89ffbc57/module/builder/consensus/builder.go#L282 Also, I found it very complicated to determine whether I had the termination condition for the traversal correct, as we have to decide whether or not to visit the _parent_ block of the given header.

In this PR I am proposing a slightly modified API for block traversing. It works for `TraverseBackward` and `TraverseForward` alike and allows to directly specify different termination conditions for including or excluding the lowest block (specifically see the last parameter):
* https://github.com/onflow/flow-go/blob/16d0eb974f4aeaa98022e11d71c612b5259b4c92/module/builder/consensus/builder.go#L436
* https://github.com/onflow/flow-go/blob/16d0eb974f4aeaa98022e11d71c612b5259b4c92/module/builder/consensus/builder.go#L327
* https://github.com/onflow/flow-go/blob/16d0eb974f4aeaa98022e11d71c612b5259b4c92/module/builder/consensus/builder.go#L199
* `fork.ExcludingHeight(lowestHeight uint64)` (not used so far)

The central change is that the conditions are all internally translated to height. Furthermore, the code also handles the case where there are no blocks to traverse. The updated code absorbs a lot of complexity into the fork traversal implementation. I find this favourable, as we can thoroughly test the logic in one place (as opposed to having to test the correct handling for each place where we use the traversal). 

Please let me know what you think. If you like the revised API for block traversal, I am happy to update the remaining usages (so far, I have left the existing implementation untouched and just added the new implementation in the separate package `fork`).

### CODE REVIEW

File: state/fork/traverse.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	return nil
+}
+
+// functor that will be called on each block header when traversing blocks.
+type sanityCheckLowestVisitedBlock = func(header *flow.Header) error
+
+// Terminal specifies the terminal condition for traversing a fork.
+// Any condition that can be converted to a block height can be
+// represented as a terminal.
+type Terminal interface {
```
Comment (durkmurder) [inline state/fork/traverse.go:110]: do we actually need an interface with one function? Maybe we can get around with functor?

**Review by durkmurder — APPROVED** [review]

**Comment by durkmurder** [comment]
Looks good to me, very clean. Generally speaking I like a lot new traversing API, would vote for replacing old one as soon as possible.

**Comment by zhangchiqing** [comment]
>Initially, I used the state.TraverseBackward and state.TraverseForward methods for the builder and seal validator. But I was running into some very subtle bugs. The main problem is that the state.TraverseBackward always visits the provided start block. However, this is undesired in case we are building on the root block:

Would it be easier to just wrap the `onVisitBlock` with a check to skip the lowest block?


```
isSealedBlock :=  func(header *flow.Header) bool {
  return header.ID() == sealedBlockID
}

state.TraverseBackward(headers, startBlockID, func(header *flow.Header) error {
  if isSealedBlock(header) {  
    return nil
  }
  return visitor(header)
},isSealedBlock)
```

File: state/fork/terminals.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+type terminalBlock struct {
+	terminalBlockID flow.Identifier
+	inclusive       bool
+}
+
+func (t terminalBlock) Translate2Height(headers storage.Headers) (uint64, sanityCheckLowestVisitedBlock, error) {
+	terminalHeader, err := headers.ByBlockID(t.terminalBlockID)
+	if err != nil {
+		return 0, noopSanityChecker, fmt.Errorf("failed to retrieve header of terminal block %x: %w", t.terminalBlockID, err)
+	}
```
Comment (jordanschalm) [inline state/fork/terminals.go:27]:
```suggestion
func (t terminalBlock) Translate2Height(headers storage.Headers) (uint64, error) {
	terminalHeader, err := headers.ByBlockID(t.terminalBlockID)
	if err != nil {
		return 0, noopSanityChecker, fmt.Errorf("failed to retrieve header of terminal block %x: %w", t.terminalBlockID, err)
	}
	
	terminalHeaderByHeight, err := headers.ByHeight(terminalHeader.Height)
	// error check
	if terminalHeader.ID() != terminalHeaderByHeight.ID() {
		// throw error
	}
```

I think it's fine to have these kinds of sanity checks, but increasing the complexity of the API to accomodate them feels like a bridge too far for me 😅. In this case I think we can limit the scope to this method as suggested above.

In general though I feel it makes more sense to put state integrity checks like this in the codepath **where the state could be changing** in way that breaks an assumption (ie. when we insert the `height->blockID` index) rather than at all the points where we read the state.

File: state/fork/traverse.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+		block, err = headers.ByBlockID(block.ParentID)
+		if err != nil {
+			return fmt.Errorf("could not get block header %x: %w", block.ParentID, err)
+		}
+	}
+}
+
+// TraverseForward traverses the given fork (specified by block ID `forkHead`)
+// in the order of increasing height. The `terminal` defines when the traversal
+// stops. The `visitor` callback is called for each block in this segment.
```
Comment (jordanschalm) [inline state/fork/traverse.go:62]:
```suggestion
// TraverseForward traverses the given fork (specified by block ID `forkHead`)
// in the order of increasing height. The `terminal` defines when the traversal
// begins. The `visitor` callback is called for each block in this segment.
```
If we are saying the traversal is in increasing height order, the terminal is defining the beginning of the traversal. 

The more I think about the `Forward` version of this API, the more I wonder if it would be simpler to just provide an API to get a range of blocks, then let the client decide how to iterate it. The main benefit of using the callback for the `TraverseBackward` case is that we can fail fast and skip the rest of the traversal if an error occurs or we hit a dynamic halting condition, but if we're traversing forward we're paying for the whole traversal anyway.
```go
func HeightRange(from, to uint64) ([]*flow.Header, error) { ... }

func Range(from, to flow.Identifier) ([]*flow.Header, error) { ... }
```

**Review by jordanschalm — COMMENTED** [review]
Haven't finished the full diff yet, but leaving a partial review focussed on the traversal for now

File: state/fork/traverse.go (hunk 1 above)
Comment (AlexHentschel) [reply state/fork/traverse.go:110]: We could. For some reason I can't really explain, I am trying to avoid functors in API methods. I associate this with code cleanliness ... But this is a highly-personal preference. Happy to change it too. 
In this case, I don't think it is performance relevant, as the `Terminal`'s `Translate2Height` method is only called once.

Comment (AlexHentschel) [reply state/fork/traverse.go:110]: Haha, I say that I prefer interfaces; but then I have implemented:
```golang
type Terminal interface {
	Translate2Height(headers storage.Headers) (uint64, sanityCheckLowestVisitedBlock, error)
}
```
where `sanityCheckLowestVisitedBlock` is a functor: https://github.com/onflow/flow-go/blob/9986c78f732eae64994985a4ea62d311a0f17197/state/fork/traverse.go#L104-L105

... I really don't know 🤷

File: state/fork/terminals.go (hunk 2 above)
Comment (AlexHentschel) [reply state/fork/terminals.go:27]: On the one hand, I totally agree that the API would be much cleaner, if we removed the `sanityCheckLowestVisitedBlock`. On the other hand, this check covers some distinct failure cases, that would otherwise return the implementation with the wrong result (instead of error'ing). Specifically:
* Consider the case where we have two forks:
   ``` 
   A <- B <- C <- D
     ^- X <- Y <- Z 
  with height:
   10   11   12   13
   ``` 
* I now call `TraverseBackward(headers, `**`D`**`, visitor, IncludingBlock(`**X**`))`. The business logic translates block `X` to height `11` and then traverses the fork with head `D` backwards until it has reached block with height `11`. As `D` and `X` are on different forks and the traversal with end with `B` as the last block; but does _not_ include `X`. This is because the inputs are inconsistent. 
* A similar edge case of inconsistent inputs is `TraverseBackward(headers, `**`D`**`, visitor, ExcludingBlock(`**X**`))`.

Of course, it is a design option to not cover the this edge case in the framework and leave it up to the application logic:
* increases complexity of the business logic (if edge cases are covered and tested) or its safety (if edge cases are uncovered)
* increases runtime complexity: This is a specific check on the terminal block. Adding this check to the visitor would require the visitor to check if the block it is currently visiting is the terminal block (or the parent). Thereby, we essentially duplicate the evaluation of the termination criterion. 
   * removing the sanity check does not significantly impact runtime complexity: as the check is only called _once_ on the terminal block

Some part of me also wants to remove this and simplify the code. Then, I am wondering, does the current design actually impact the complexity of the logic calling it? 
https://github.com/onflow/flow-go/blob/9986c78f732eae64994985a4ea62d311a0f17197/module/builder/consensus/builder.go#L199 would still look exactly the same when removing the sanity check. 

So here is the logical summary from my perspective: 
* the sanity check makes the framework safer because it allows the framework to error in the case of inconsistent input (as opposed to returning possibly incorrect results)
* the sanity check makes increases internal complexity of the framework 
* the sanity check makes does not impact external API complexity

Comment (AlexHentschel) [reply state/fork/terminals.go:27]:
```golang
terminalHeaderByHeight, err := headers.ByHeight(terminalHeader.Height)
```

Sorry, forgot to mention: This only works for finalized blocks. What if you want to traverse unfinalized blocks (or orphaned blocks for whatever reason).

File: state/fork/traverse.go (hunk 3 above)
Comment (AlexHentschel) [reply state/fork/traverse.go]: Thanks for the documentation revision. 

Regarding the `Range` suggestion, I agree that this is nice for traversing forward. Though, in the dominant number of cases, we are actually traversing backwards. To me, this seems like we have three options:
1. use `Range` everywhere, which means we are paying the cost of unnecessarily allocating the slice `[]*flow.Header`  for traversing backwards (most of the cases)
2. use different APIs depending on direction `TraverseBackward` and `Range` 
3. use consistent API (i.e. `TraverseBackward` and `TraverseForward`) 

While I opted for 3., I totally agree that it is a matter of personal preference. Nevertheless, I think that 3. is the most complex to implement, but offers the combination of performance and API consistency.

**Comment by AlexHentschel** [comment]
> Would it be easier to just wrap the `onVisitBlock` with a check to skip the lowest block?

Yea, I have contemplated using this approach @zhangchiqing. In my mind, the block traversal functions are service functionality. So my goal here is not necessarily to provide the simplest possible implementation of the service functionality, but rather the most general and versatile one. If we have the choice to either absorb complexity into this service functionality or leave it to the higher-order logic, I would opt for the former. 

In my mind, what you are suggesting is to have the higher-order business logic implement whether to include or exclude the terminal block. Then, we would duplicate this implementation and would have to test it each time. In contrast, when it is part of the service functionality, we only need to test the logic once. The and the higher-order business logic only has to declare its intend and can rely on the implementation being correct.

File: state/fork/traverse.go (hunk 1 above)
Comment (durkmurder) [reply state/fork/traverse.go:110]: Originally I've made the comment because it's only one function and there is no sign of adding more. On other hand we will have functor that returns a tuple where one of members is another functor, so maybe the interface is actually a good choice here.

File: state/fork/traverse.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	return unsafeTraverse(headers, startBlock, visitor, lowestHeightToVisit, sanityChecker)
+}
+
+// unsafeTraverse implements the fork traversal in the order of decreasing height.
+// It is unsafe because:
+// * always calls the `visitor` on the `block`, _before_ it checks whether to stop
+// * if `block` has a lower height than lowestHeightToVisit, the traversal keeps
+//   going until we hit the root block
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply state/fork/traverse.go:37]: In fact, we could consider moving the pre-check into the `unsafeTraverse`, and remove the `unsafe` in the name. This could work, because unsafeTraverse takes `startBlock` and `lowestheightToVisit` anyway

File: state/fork/traverse.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+func unsafeTraverse(headers storage.Headers, block *flow.Header, visitor onVisitBlock, lowestHeightToVisit uint64, sanityChecker sanityCheckLowestVisitedBlock) error {
+	for {
+		err := visitor(block)
+		if err != nil {
+			return err
+		}
+
+		if block.Height == lowestHeightToVisit {
+			err = sanityChecker(block)
```
Comment (zhangchiqing) [inline state/fork/traverse.go:46]: I see the point of having `sanityChecker`, but the `unsafeTraverse` doesn't really need this sanityCheck, having the `unsafeTraverse` taking `sanityChecker` as parameter increases the complexity.

I would suggest to let `unsafeTraverse` return the last visited block, and then the `TraverseForward` and `TraverseBack` could use the sanityChecker to check the last block.

This makes `unsafeTraverse` simplier, and I believe could make `TraverseBackard` easier to understand.

Comment (zhangchiqing) [reply state/fork/traverse.go:46]: another issue is I found the name `sanityCheck` is confusing. I think this is one of the source of increased complexity. 
The question I had when seeing this are: Why do we need sanityCheck? What are these checks? 
I have to navigate to the `Terminal` implementation to understand. 

After reading the sanityChecker, I learned it's to check if all visited blocks are on the same chain. This is a nice check, because if the caller specifies a fork header block and a terminal block to traverse, but what if these two blocks are not on the same chain? It might risk running into an endless loop. 

And this sanity check is only useful if the terminal block is specified by block ID. If it's specified by block height, then there is nothing to sanity check, because we essentially just traverse certain blocks along the chain.

So one suggestion is to just rename `sanityCheck` to `sameChainCheck` to be more specific.

File: state/fork/traverse_test.go [hunk 6]
```diff
@@ -0,0 +1,543 @@
+package fork
```
Comment (zhangchiqing) [inline state/fork/traverse_test.go:1]: Great tests 💯

**Review by zhangchiqing — APPROVED** [review]
The change to the sealing logic looks good. 

Had some suggestion to the traverse function. And thanks for putting the effort on the tests, they are great 👏
